src/base/bittorrent/torrenthandle.cpp

```diff
@@ -578,6 +594,48 @@ bool TorrentHandle::belongsToCategory(const QString &category) const
     return false;
 }
 
+const QSet<QString> &TorrentHandle::tags() const
+{
+    return m_tags;
+}
+
+bool TorrentHandle::hasTag(const QString &tag) const
+{
+    if (!Session::isValidTag(tag))
+        return false;
+    return m_tags.contains(tag);
+}
+
+bool TorrentHandle::addTag(const QString &tag)
+{
+    if (Session::isValidTag(tag) && !hasTag(tag)) {
+        if (!m_session->hasTag(tag) && !m_session->addTag(tag))
```

Comment:
It needs to be separate because there are 3 different cases here:
- Both torrent and session have the tag already (do nothing).
- The session has the tag but the torrent doesn't (add the tag to the torrent).
- Neither the session nor the torrent has the tag (add the tag to both of them).

...and I have actually split these conditionals up further at the request of another review comment.